Index: madwifi-dfs-r3319-20080201/ath/if_ath.c
===================================================================
--- madwifi-dfs-r3319-20080201.orig/ath/if_ath.c
+++ madwifi-dfs-r3319-20080201/ath/if_ath.c
@@ -8506,8 +8506,6 @@
 	ath_hal_intrset(sc->sc_ah, sc->sc_imask);
 	local_irq_restore(flags);
 
-	netif_wake_queue(dev);
-
 	if (sc->sc_softled)
 		ath_led_event(sc, ATH_LED_TX);
 }
@@ -8554,8 +8552,6 @@
 	ath_hal_intrset(sc->sc_ah, sc->sc_imask);
 	local_irq_restore(flags);
 
-	netif_wake_queue(dev);
-
 	if (sc->sc_softled)
 		ath_led_event(sc, ATH_LED_TX);
 }
@@ -8588,8 +8584,6 @@
 	ath_hal_intrset(sc->sc_ah, sc->sc_imask);
 	local_irq_restore(flags);
 
-	netif_wake_queue(dev);
-
 	if (sc->sc_softled)
 		ath_led_event(sc, ATH_LED_TX);
 }
Index: madwifi-dfs-r3319-20080201/net80211/ieee80211_input.c
===================================================================
--- madwifi-dfs-r3319-20080201.orig/net80211/ieee80211_input.c
+++ madwifi-dfs-r3319-20080201/net80211/ieee80211_input.c
@@ -1141,7 +1141,7 @@
 	    (vap->iv_flags & IEEE80211_F_NOBRIDGE) == 0) {
 		struct sk_buff *skb1 = NULL;
 
-		if (ETHER_IS_MULTICAST(eh->ether_dhost)) {
+		if (ETHER_IS_MULTICAST(eh->ether_dhost) && !netif_queue_stopped(dev)) {
 			skb1 = skb_copy(skb, GFP_ATOMIC);
 			/* We duplicate the reference after skb_copy */
 			ieee80211_skb_copy_noderef(skb, skb1);
